### PR TITLE
Add optional predicate to Single and SingleOrDefault

### LINKED ISSUE #82: SingleOrDefault doesnt accept predicate?
In .net Linq you can do `myList.SingleOrDefault(x => x.someId == requiredId);` but I noticed in here there is no overload to support this like in .net.

https://msdn.microsoft.com/en-us/library/bb549274(v=vs.110).aspx

### PR DESCRIPTION
**Pull request by grofit**
Added support for optional predicate with `Single` and `SingleOrDefault`.

Fixes #82

### CODE REVIEW

File: test/list.ts [hunk 1]
```diff
@@ -683,24 +727,44 @@ test('Single', t => {
     () => fruits3.Single(),
     /The collection does not contain exactly one element./
   )
+  t.is(numbers1.Single(x => x == 1), 1)
```
Comment (kutyel) [inline test/list.ts]: Please always use triple equals (`===`), after all, this is JavaScript :)

**Comment by kutyel** [comment]
Awesome! Thanks very much, just replace `==` with `===` and everything will be perfect for merging! 👏 
BTW: you can also support the library by givig it a ⭐️  on GitHub! 😉

File: test/list.ts (hunk 1 above)
Comment (grofit) [reply test/list.ts]: I find it odd you say that when all the method names are `UpperCamelCase/PascalCase` and not `lowerCamelCase` ;) Anyway === added.

**Comment by kutyel** [comment]
Awesome work, thanks! 🎉  How did you find the codebase? Was it easy to contribute? :)

**Comment by grofit** [comment]
Yeah fine, only odd thing as I said was that use have `PascalCase` where most js libs these days are `lowerCamelCase`. I wont lost any sleep over it though, also that its `new List<T>` rather than `new Enumerable<T>` or something to make it more explicitly linq like, but it does the job, and thats all I need as most linq implementations in js have out of date docs/d.ts files.

**Pull request merged**